File: src/litcoder/module-4/lab-2/EgyptianFraction.py
"""Solution to Egyptian Fraction  problem"""
import math
from typing import List


def solution(nr: int, dr: int) -> List[int]:
    """Returns the denominators in egyptian fraction

    Args:
        nr (int): numerator
        dr (int): denominator

    Returns:
        List[int]: list of denominators
    """
    denominators: List[int] = []

    # run loop until nr becomes 0
    while nr != 0:
        x = math.ceil(dr / nr)

        # storing value in ef list
        denominators.append(x)

        # update nr and dr
        nr = x * nr - dr
        dr = dr * x

    return denominators


testcases = int(input())
for _ in range(testcases):
    numerator, denominator = map(int, input().split(" "))
    print(solution(numerator, denominator))

File: src/codes/college/litcoder/module4/lab2/EgyptianFraction.java
package codes.college.litcoder.module4.lab2;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class EgyptianFraction {

    /**
     * Calculates the Egyptian fraction representation of a given fraction.
     * An Egyptian fraction is a sum of distinct unit fractions, where a unit
     * fraction is a fraction with numerator 1.
     * The function takes the numerator and denominator of the fraction as input and
     * returns a list of denominators
     * representing the Egyptian fraction.
     *
     * @param nr the numerator of the fraction
     * @param dr the denominator of the fraction
     * @return a list of denominators representing the Egyptian fraction
     */
    public static List<Integer> solution(int nr, int dr) {
        List<Integer> denominators = new ArrayList<>();

        // Run the loop until nr becomes 0
        while (nr != 0) {
            int x = (int) Math.ceil((double) dr / nr);

            // Storing value in the ef list
            denominators.add(x);

            // Update nr and dr
            nr = x * nr - dr;
            dr = dr * x;
        }

        return denominators;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        int numerator = scanner.nextInt();
        int denominator = scanner.nextInt();

        List<Integer> result = solution(numerator, denominator);
        for (int i : result) {
            System.out.println(i);
        }

        scanner.close();
    }
}
